### PR TITLE
adds the linter public_member_api_docs and adds docs where missing

### PR DESCRIPTION
**Pull request by frankpepermans**
From the title:
- adds the new linter public_member_api_docs
- adds docs where the linter complained

In 90% of the cases, we were not documenting the class constructor(s).
I've simply added a 'readers digest' derived from the main class documentation above in most cases, with some different wording, and [links] to ctr arguments if applicable.

In other cases, simply making public properties private was required,
for example some Streams use a controller, which is final and passed in the constructor arguments.

This controller can simply be a private _controller, we don't want this to be accessed outside the class anyway...

### CODE REVIEW

**Comment by frankpepermans** [comment]
Ah, I made them private so that they cannot be fiddled with down the line.

As for example a List of something could be just final and public, but you can still do list.add if it is accessible.

Maybe we could drop the underscore, but mark the property with @private then?

File: lib/src/streams/concat.dart [hunk 1]
```diff
@@ -17,15 +17,20 @@ import 'dart:async';
 ///     ])
 ///     .listen(print); // prints 1, 2, 3
 class ConcatStream<T> extends Stream<T> {
-  final StreamController<T> controller;
+  final StreamController<T> _controller;
 
+  /// Constructs a [Stream] which emits all events from [streams].
+  /// The [Iterable] is traversed upwards, meaning that the current first
+  /// [Stream] in the [Iterable] needs to complete, before events from the
+  /// next [Stream] will be emitted.
```
[the remark this answers is not in the document]
Comment (frankpepermans) [reply lib/src/streams/concat.dart]: Indeed, fixing that

File: lib/src/streams/retry_when.dart [hunk 2]
```diff
@@ -60,13 +60,17 @@ import 'package:rxdart/src/streams/utils.dart';
...
+  StreamController<T> _controller;
+  StreamSubscription<T> _subscription;
   List<ErrorAndStacktrace> _errors = <ErrorAndStacktrace>[];
 
-  RetryWhenStream(this.streamFactory, this.retryWhenFactory);
+  /// Constructs a [Stream] that will recreate and re-listen to the source
+  /// [Stream] (created by the provided factory method).
+  /// The retry will trigger whenever the [Stream] created by the retryWhen
+  /// factory emits and event.
+  RetryWhenStream(this._streamFactory, this._retryWhenFactory);
```
[the remark this answers is not in the document]
Comment (frankpepermans) [reply lib/src/streams/retry_when.dart]: We build the `_controller` always right after the` constructor` is called in almost all other classes.

Here, it is being built on `bind()`,
I've decided to keep consistency instead, where `streamFactory` and `retryWhenFactory` are not kept as class parameters, but instead directly passed to `_buildController`.

This way the properties in the constructor are just local properties.

File: lib/src/transformers/group_by.dart [hunk 3]
```diff
@@ -9,19 +9,21 @@ import 'package:rxdart/src/observables/observable.dart' show Observable;
...
 
 class GroupByStreamTransformer<T, S>
     extends StreamTransformerBase<T, GroupByObservable<T, S>> {
-  final S Function(T) grouper;
+  final S Function(T) _grouper;
 
-  GroupByStreamTransformer(this.grouper);
+  /// Constructs a [StreamTransformer] which groups events from the source
+  /// [Stream] and emits them as [GroupByObservable].
+  GroupByStreamTransformer(this._grouper);
```
[the remark this answers is not in the document]
Comment (frankpepermans) [reply lib/src/transformers/group_by.dart]: See above

File: lib/src/transformers/scan.dart [hunk 4]
```diff
@@ -10,14 +10,17 @@ import 'dart:async';
...
-  final _ScanStreamTransformerAccumulator<T, S> accumulator;
-  final S seed;
+  final _ScanStreamTransformerAccumulator<T, S> _accumulator;
+  final S _seed;
 
-  ScanStreamTransformer(this.accumulator, [this.seed]);
+  /// Constructs a [ScanStreamTransformer] which applies an accumulator Function
+  /// over the source [Stream] and returns each intermediate result.
+  /// The optional seed value is used as the initial accumulator value.
+  ScanStreamTransformer(this._accumulator, [this._seed]);
```
[the remark this answers is not in the document]
Comment (frankpepermans) [reply lib/src/transformers/scan.dart]: Same as above

**Comment by frankpepermans** [comment]
So to recap,

In general we don't keep constructor params inside the Stream/Transformer class scope, usually we do:

`SomeConstructor(Foo foo, Bar bar) : _controller = _buildController(foo, bar);`

However, in a few (older) classes, we did:

```
SomeConstructor(this.foo, this.bar);

@override
Stream bind(stream) =>  _buildController(foo, bar).bind(stream);
```
If we want to keep constructor params as public class members instead, then we'd need to change all of our implementations.

Personally I wouldn't keep these params around, they are no longer needed after the `controller` is built (for us) and I think they would only clutter the api.

**Comment by brianegan** [comment]
Ah, good call! In that case, I think this is looking really good. Nice one :)

**Comment by frankpepermans** [comment]
@brianegan OK to merge?

I know you're working on the extension method branch, if the new linter is in place, your branch will need to copy most of these new docs?

**Comment by brianegan** [comment]
Yep, ready to merge!

To pull these changes in, I'm hoping I can just do a rebase. I can only see 1-2 conflicts arising since this generally affects the classes directly.

**Pull request merged**